### PR TITLE
Add search feature

### PR DESCRIPTION
**Pull request by danielbrzn**
Resolves https://github.com/MarkBind/markbind/issues/151

This PR adds a search feature to Markbind that can be used to traverse the titles in site.json and route the user to the appropriate page. It uses the Typeahead component available in VueStrap.

Desktop:
image

Mobile:
image

It can be added to any page using the following HTML code:
```
<typeahead :data="searchData" placeholder="Search" :on-hit="searchCallback"></typeahead>
```

To add it onto the navbar, wrap the above HTML code in a form as follows:

```
<form class="navbar-form">
  <typeahead :data="searchData" placeholder="Search" :on-hit="searchCallback"></typeahead>
</form>
```

### Additional considerations:
- ~~Should there be parity between mobile and desktop for the highlight colour of the items?~~

### To-do:
- ~~Ensure that dropdown text on mobile wraps around so that the user can see the whole title.~~

### CODE REVIEW

File: asset/js/setup.js [hunk 1]
```diff
@@ -1,7 +1,36 @@
...
+    data() {
+      return {
+        searchData: titleArray,
+      };
+    },
+    methods: {
+      searchCallback(match) {
+        const index = titleArray.indexOf(match);
+        const route = routeArray[index];
+        window.location.pathname = route.replace('md', 'html');
```
Comment (nicholaschuayunzhi) [inline asset/js/setup.js]: If the route contains `md` in the name, it will give an incorrect route.

File: asset/js/setup.js [hunk 2]
```diff
@@ -1,7 +1,36 @@
 /* eslint-disable no-undef */
 
 Vue.use(VueStrap);
-const vm = new Vue({
-  el: '#app',
+
+const typeahead = VueStrap.components.typeahead;
+
+jQuery.getJSON('../../site.json').then((siteData) => {
```
Comment (nicholaschuayunzhi) [inline asset/js/setup.js]: Can remove `(` in arrow function

File: asset/js/setup.js (hunk 1 above)
Comment (rachx) [reply asset/js/setup.js]: Using the path module, 
1. Check if `path.extname(route)` is `.md`
2. `path.join(path.dirname(route), path.basename(route, path.extname) + '.html')`

File: asset/js/setup.js (hunk 2 above)
Comment (nicholaschuayunzhi) [reply asset/js/setup.js]: Ignore this. The rule requires brackets if function body is wrapped by curly brace

File: asset/js/setup.js (hunk 1 above)
Comment (danielbrzn) [reply asset/js/setup.js]: Thanks! Will change it to replace the extensions instead of just 'md'.

**Comment by acjh** [comment]
Tried both the HTML in PR description with [nus-cs2103/website-base](https://github.com/nus-cs2103/website-base), but I don't see the search bar.

**Comment by danielbrzn** [comment]
Are there any errors in the chrome dev console?

Just tried it on my end but unable to repro.

**Comment by acjh** [comment]
Whoops, I was on `search-bar` branch.

> ```html
> <form class="navbar-form">
>   <typeahead :data="searchData" placeholder="Search" :on-hit="searchCallback"></typeahead>
> </form>
> ```

What's the purpose of the `form` element, and how did you get it to align to the right of the navbar?

**Comment by danielbrzn** [comment]
The `navbar-form` class lets the search bar be vertically centered within the navbar, wherever it's placed. Also gives it appropriate padding/margin.

Without:
image

With:
image

To get it to align to the right of the navbar, you can use the slot attribute for a list.

```html
<li slot="right">
  <form class="navbar-form">
    <typeahead :data="searchData" placeholder="Search" :on-hit="searchCallback"></typeahead> 
  </form>
</li>
```

File: asset/js/setup.js [hunk 3]
```diff
@@ -1,7 +1,36 @@
 /* eslint-disable no-undef */
 
 Vue.use(VueStrap);
-const vm = new Vue({
-  el: '#app',
+
+const typeahead = VueStrap.components.typeahead;
```
Comment (acjh) [inline asset/js/setup.js]: ESLint:

```js
const { typeahead } = VueStrap.components.typeahead;
```

File: asset/js/setup.js (hunk 2 above)
Comment (acjh) [inline asset/js/setup.js]: Let's abstract this:

```js
jQuery.getJSON('../../site.json')
  .then(setupWithSearch)
  .catch(setup);
```

File: asset/js/setup.js [hunk 4]
```diff
@@ -1,7 +1,41 @@
 /* eslint-disable no-undef */
 
 Vue.use(VueStrap);
-const vm = new Vue({
-  el: '#app',
-});
-VueStrap.installEvents(vm);
+
+const { typeahead } = VueStrap.components;
```
Comment (acjh) [inline asset/js/setup.js]: Can this be in `setupWithSearch`?

File: asset/js/setup.js [hunk 5]
```diff
@@ -1,7 +1,41 @@
...
+    data() {
+      return {
+        searchData: titleArray,
+      };
+    },
+    methods: {
+      searchCallback(match) {
+        const index = titleArray.indexOf(match);
+        const route = routeArray[index];
+        window.location.pathname = route.replace('.md', '.html');
```
Comment (acjh) [inline asset/js/setup.js:30]: The change in https://github.com/MarkBind/markbind-cli/pull/32#discussion_r170452815 wasn't made?

File: asset/js/setup.js [hunk 6]
```diff
@@ -1,7 +1,41 @@
...
+  const vm = new Vue({
+    el: '#app',
+  });
+  VueStrap.installEvents(vm);
+}
+
+function setupWithSearch(siteData) {
+  const titleArray = jQuery.map(siteData.pages, object => object.title);
+  const routeArray = jQuery.map(siteData.pages, object => object.src);
+
```
Comment (acjh) [inline asset/js/setup.js]: - Alphabetical order
- Remove newline

File: asset/js/setup.js (hunk 5 above)
Comment (danielbrzn) [reply asset/js/setup.js:30]: It was. I changed it to replace '.md' with '.html' instead of just 'md'.

**Comment by acjh** [comment]
*Remember to comment when you update a PR with a force-push because GitHub doesn't send notification emails for that.*

**Review by acjh — APPROVED** [review]

**Comment by Gisonrg** [comment]
I think the `typeahead` name is not intuitive :P Maybe we should change it to more readable name like "searchBox". But this PR should be ready to go 👍

**Comment by damithc** [comment]
I used the following to add a search box to the nav bar. It's showing up, but not giving any results. Any ideas? A restart of the server (in case search data are generated on restart) didn't help.
```html
  <li slot="right">
    <form class="navbar-form">
      <typeahead :data="searchData" placeholder="Search" :on-hit="searchCallback"></typeahead> 
    </form>
  </li>
```
image

**Comment by danielbrzn** [comment]
Is site.json being ignored in the site.json file?

**Comment by damithc** [comment]
> Is site.json being ignored in the site.json file?

yes, that was the problem. Working now 👍

**Pull request merged**